lib/Onyx.js

```diff
@@ -978,16 +972,16 @@ function applyMerge(key, data) {
  * Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Workspace'}
  *
  * @param {String} key ONYXKEYS key
- * @param {(Object|Array)} value Object or Array value to merge
+ * @param {(Object|Array)} changes Object or Array value to merge
  * @returns {Promise}
  */
-function merge(key, value) {
+function merge(key, changes) {
     if (mergeQueue[key]) {
-        mergeQueue[key].push(value);
+        mergeQueue[key].push(changes);
         return Promise.resolve();
     }
```

Comment:
To avoid race conditions something like the following might be used (using the SyncQueue would be similar, but blocking for all keys)

```js
// pending merges by key
const mergePromises = {};

function merge(key, changes) {
  // wait any prior updates by chaining after an existing promise or set up a new chain
  const megePromise = mergePromises[key] || Promise.resolve();

  mergePromises[key] = mergePromise
    .then(() => get(key))
    .then(data => {
      const modifiedData = applyMerge(key, data);      

      // We relay changes to subscribers optimistically to avoid lag (think UI updates)
      notify(key, modifiedData, 'merge');
      return Storage.merge(key, changes, modifiedData);
    })
    .catch((error) => Logger.logAlert(`An error occurred while applying merge for key: ${key}, Error: ${error}`));

  return mergePromises[key];
}
```

Reply:
I like that idea! Let's do that

Reply:
With this approach, we'll still need to explicitly wait for a merge to complete, for example if we call `Onyx.clear`/`Onyx.set` right after...

Ofc, multiple merges will wait for the the previous merge to complete, but we would also need to add waiting to any other Onyx functions

Reply:
I think keys are only used in a certain way - we either update a key using only `merge` or only `set` - they are rarely mixed 
It's unlikely that a merge and a set for the same key would overlap - updates are primarily merged 
The initial value might be defined with set though I'm not sure 

It's true that nothing should run after Onyx.clear, but what do we do about it right now - I think it's covered by existing logic, and it would remain covered by the same logic

Reply:
Ok, got it. We'd still need to update tests, so those cover only the real scenarios, that we support. Because atm there are also some tests, where set and merge are used successively

Reply:
I'm mostly referring to [this thread on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1683300135739449)

Reply:
To clarify: my proposal is to promisify the `merge` func so it waits any running merge for the same key: https://github.com/Expensify/react-native-onyx/pull/238#discussion_r1184585616

We would also like to `logAlert` when either `set` is called for a key we're still merging or vice versa
I wouldn't worry much about clear, because it very unlikely to overlap with a merge and IIRC calling `clear` aborts any pending merges in the queue

Back to `logAlert` - to make this work we'll need to temporarily store which key is currently being set/merged 
Instead of doing that we can just as easily put a promise in `mergePromises` (let's rename that to `keyOperations` or `keyWrites`) 
Now whenever something is being written to a key, other methods that want to write to the same key can wait

```js
// pending set or merges by key
const keyOperations = {};

function merge(key, changes) {
  // wait any prior updates by chaining after an existing promise or set up a new chain
  const writePromise = keyOperations[key] || Promise.resolve();

  keyOperations[key] = writePromise
    .then(() => get(key))
    .then(data => { ... })
    .catch((error) => Logger.logAlert(`An error occurred while applying merge for key: ${key}, Error: ${error}`));

  return keyOperations[key];
}
```
`function set` would be doing something similar 

```js
function set(key, value) {
  // ... existing logic

  // wait any prior updates by chaining after an existing promise or set up a new chain
  const writePromise = keyOperations[key] || Promise.resolve();

   keyOperations[key]  = writePromise
     .then(() => Storage.setItem(key, value))
     .catch(error => evictStorageAndRetry(error, set, key, value))

  return keyOperations[key];
}
```

Adding logging would have brought similar logic in order to track overlaps, but we can just solve the problem, by making key writes wait each other (when they overlap)

The same concept can be applied to `function remove` so it locks the `key` by setting a promise in `keyOperations[key]` - any writes that started after `remove()` would wait and proceed after

---

Whatever we do we also need to find a way to verify any changes to `function merge` did not introduce regressions

Reply:
Thanks, that makes sense!

Reply:
`Onyx.clear` is not aborting current merge's, so we can go for one of these 3 options:

1. Use the same concept (as above) in `clear` and wait for `merge`/`set`/`remove` to be done, before clearing
2. Abort current promise, e.g. with an `AbortController`
3. Ignore this case, and remove tests that therefore currently fail, such as this one:

```js
    it('should replace the value of Onyx.merge with the default key state in the cache', () => {
        expect.assertions(3);

        // Given that Onyx is completely clear
        // When merge then clear is called on a key with a default key state
        Onyx.merge(ONYX_KEYS.DEFAULT_KEY, MERGED_VALUE);
        Onyx.clear();
        return waitForPromisesToResolve()
            .then(() => {
                // Then the value in Onyx, the cache, and Storage is the default key state
                expect(onyxValue).toBe(DEFAULT_VALUE);
                const cachedValue = cache.getValue(ONYX_KEYS.DEFAULT_KEY);
                expect(cachedValue).toBe(DEFAULT_VALUE);
                const storedValue = StorageMock.getItem(ONYX_KEYS.DEFAULT_KEY);
                return expect(storedValue).resolves.toBe(DEFAULT_VALUE);
            });
    });
```

cc @kidroca

Reply:
Since `clear` does not wait `merge` I find it interesting how this test was passing before?

This is what I meant, when I said `clear` aborts write operations:
https://github.com/Expensify/react-native-onyx/blob/1d496b001f327c509131597faf260a81c002be77/lib/storage/providers/LocalForage.js#L84-L87

I think that was only needed on Web, because write operations are queued

I think you can make `clear` wait all the `keyOperations` and then run, but I don't think this would be necessary for App to function correctly
Aborting would be optimal, but it would introduce more complexity